engine/access/rest/websockets/controller.go

```diff
@@ -111,214 +113,277 @@ func (c *Controller) configureKeepalive() error {
 	return nil
 }
 
+// keepalive sends a ping message periodically to keep the WebSocket connection alive
+// and avoid timeouts.
+//
+// Expected errors during normal operation:
+// - context.Canceled if the client disconnected
+func (c *Controller) keepalive(ctx context.Context) error {
+	pingTicker := time.NewTicker(PingPeriod)
+	defer pingTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-pingTicker.C:
+			err := c.conn.WriteControl(websocket.PingMessage, time.Now().Add(WriteWait))
+			if err != nil {
+				if errors.Is(err, websocket.ErrCloseSent) {
+					return err
```

Comment:
is this an error condition?

Reply:
It is not but we propagate every ErrCloseSent to `HandleConnection()`. `HandleConnection()` decides what to do with it. (currently nothing)